internal/testing/oteltest/exporter.go

```diff
@@ -0,0 +1,236 @@
+package oteltest
+
+import (
+	"context"
+	iotel "gocloud.dev/internal/otel"
+	"sync"
+	"testing"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	noopmetric "go.opentelemetry.io/otel/metric/noop"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+	nooptrace "go.opentelemetry.io/otel/trace/noop"
+)
+
+// TestExporter is an exporter of OpenTelemetry traces and metrics, for testing.
+// It should be created with NewTestExporter.
+type TestExporter struct {
+	mu             sync.Mutex
+	spanExporter   *tracetest.InMemoryExporter
+	metricExporter *MetricExporter
+	shutdown       func(context.Context) error
+}
+
+// MetricExporter is a simple metrics exporter for testing.
+type MetricExporter struct {
+	mu       sync.Mutex
+	metrics  []metricdata.ScopeMetrics
+	metricCh chan metricdata.ScopeMetrics
+	reader   *sdkmetric.PeriodicReader
+}
+
+// NewMetricExporter creates a new metric exporter for testing
+func NewMetricExporter() *MetricExporter {
+	exporter := &MetricExporter{
+		metricCh: make(chan metricdata.ScopeMetrics, 10),
+	}
+
+	exporter.reader = sdkmetric.NewPeriodicReader(
+		exporter,
+		// Use a short export interval for tests
+		sdkmetric.WithInterval(100*time.Millisecond),
+	)
+
+	return exporter
+}
+
+// Temporality returns the aggregation temporality for the given instrument kind.
+func (e *MetricExporter) Temporality(kind sdkmetric.InstrumentKind) metricdata.Temporality {
+	return metricdata.CumulativeTemporality
+}
+
+// Aggregation returns the aggregation for the given instrument kind.
+func (e *MetricExporter) Aggregation(kind sdkmetric.InstrumentKind) sdkmetric.Aggregation {
+	return nil // Use default aggregation
+}
+
+// Export exports metric data.
+func (e *MetricExporter) Export(ctx context.Context, data *metricdata.ResourceMetrics) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	// Store metrics for each scope
+	for _, sm := range data.ScopeMetrics {
+		e.metrics = append(e.metrics, sm)
+
+		// Also send via channel
+		select {
+		case e.metricCh <- sm:
+		default:
+		}
+	}
+
+	return nil
+}
+
+// ForceFlush forces a flush of metrics
+func (e *MetricExporter) ForceFlush(ctx context.Context) error {
+	return nil
+}
+
+// GetMetrics returns all collected metrics
+func (e *MetricExporter) GetMetrics() []metricdata.ScopeMetrics {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	return e.metrics
+}
+
+// WaitForMetrics waits for metrics to be collected
+func (e *MetricExporter) WaitForMetrics(timeout time.Duration) (metricdata.ScopeMetrics, bool) {
+	select {
+	case sm := <-e.metricCh:
+		return sm, true
+	case <-time.After(timeout):
+		return metricdata.ScopeMetrics{}, false
+	}
+}
+
+// Shutdown shuts down the exporter
+func (e *MetricExporter) Shutdown(ctx context.Context) error {
+	return e.reader.Shutdown(ctx)
```

Comment:
I am pretty sure this loops forever, because reader calls exporter.Shutdown:

https://github.com/open-telemetry/opentelemetry-go/blob/69e81088ad40f45a0764597326722dea8f3f00a8/sdk/metric/periodic_reader.go#L337

Reply:
well noted, thank you for being eagle eyed. I will just reset the metrics exporter